packages/astro/src/vite-plugin-astro-server/plugin.ts

```diff
@@ -50,26 +53,46 @@ export default function createVitePluginAstroServer({
 			const controller = createController({ loader });
 			const localStorage = new AsyncLocalStorage();
 
-			/** rebuild the route cache + manifest, as needed. */
-			async function rebuildManifest(needsManifestRebuild: boolean) {
+			/** rebuild the route cache + manifest */
+			async function rebuildManifest(path: string | null = null) {
 				pipeline.clearRouteCache();
-				if (needsManifestRebuild) {
+
+				// If a route changes, we check if it's part of the manifest and check for its prerender value
+				if (path !== null) {
+					const route = routeManifest.routes.find(
+						(r) => path === new URL(r.component, settings.config.root).pathname,
+					);
+					if (!route) {
+						return;
+					}
+					if (route.type !== 'page' && route.type !== 'endpoint') return;
+
+					const routePath = fileURLToPath(new URL(route.component, settings.config.root));
+					if (!fsMod.existsSync(routePath)) {
+						// Route was renamed so it does not exist anymore
+						return;
+					}
+					const content = await fsMod.promises.readFile(routePath, 'utf-8');
+					await getRoutePrerenderOption(content, route, settings, logger);
+				} else {
 					routeManifest = injectDefaultDevRoutes(
 						settings,
 						devSSRManifest,
-						await createRouteManifest({ settings, fsMod }, logger), // TODO: Handle partial updates to the manifest
+						await createRouteManifest({ settings, fsMod }, logger, { dev: true }),
 					);
-					warnMissingAdapter(logger, settings);
-					pipeline.manifest.checkOrigin =
-						settings.config.security.checkOrigin && settings.buildOutput === 'server';
-					pipeline.setManifestData(routeManifest);
 				}
+				await runHookRoutesResolved({ routes: routeManifest.routes, settings, logger });
+
+				warnMissingAdapter(logger, settings);
+				pipeline.manifest.checkOrigin =
+					settings.config.security.checkOrigin && settings.buildOutput === 'server';
+				pipeline.setManifestData(routeManifest);
 			}
 
-			// Rebuild route manifest on file change, if needed.
-			viteServer.watcher.on('add', rebuildManifest.bind(null, true));
-			viteServer.watcher.on('unlink', rebuildManifest.bind(null, true));
-			viteServer.watcher.on('change', rebuildManifest.bind(null, false));
+			// Rebuild route manifest on file change
+			viteServer.watcher.on('add', () => rebuildManifest());
+			viteServer.watcher.on('unlink', () => rebuildManifest());
+			viteServer.watcher.on('change', (path) => rebuildManifest(path));
```

Comment:
```suggestion
			viteServer.watcher.on('add', rebuildManifest);
			viteServer.watcher.on('unlink', rebuildManifest);
			viteServer.watcher.on('change', rebuildManifest);
```

I think this should work